--- a/src/mixer/mixer.h
+++ b/src/mixer/mixer.h
@@ -30,0 +31 @@
+  ContextData* active_data_;
--- a/src/mixer/mixer.cpp
+++ b/src/mixer/mixer.cpp
@@ -16 +16,2 @@
-    learning_rate_(learning_rate), context_(context), max_steps_(1), steps_(0)
+    learning_rate_(learning_rate), active_data_(nullptr), context_(context),
+    max_steps_(1), steps_(0)
@@ -42,0 +44 @@
+  active_data_ = data;
@@ -74 +76 @@
-  ContextData* data = GetContextData();
+  ContextData* data = active_data_;
